proposals/0540-jsem.rst

```diff
@@ -0,0 +1,384 @@
+~~~~~~~~~~~~~~~~~~~~~~
+
+This proposal introduces the GHC Jobserver Protocol. This protocol allows
+a server to dynamically invoke many instances of a client process,
+while restricting all of those instances to use no more than <n> capabilities.
+This is achieved by coordination over a system semaphore (either a POSIX
+semaphore [6]_  in the case of Linux and Darwin, or a Win32 semaphore [7]_
+in the case of Windows platforms).
+
+There are two kinds of participants in the GHC Jobserver protocol:
+
+- The *jobserver* creates a system semaphore with a certain number of
+  available tokens.
+
+  Each time the jobserver wants to spawn a new jobclient subprocess, it **must**
+  first acquire a single token from the semaphore, before spawning
+  the subprocess. This token **must** be released once the subprocess terminates.
+
+  Once work is finished, the jobserver **must** destroy the semaphore it created.
+
+- A *jobclient* is a subprocess spawned by the jobserver or another jobclient.
+
+  Each jobclient starts with one available token (its *implicit token*,
+  which was acquired by the parent which spawned it), and can request more
+  tokens through the Jobserver Protocol by waiting on the semaphore.
+
+  Each time a jobclient wants to spawn a new jobclient subprocess, it **must**
+  pass on a single token to the child jobclient. This token can either be the
+  jobclient's implicit token, or another token which the jobclient acquired
+  from the semaphore.
+
+  Each jobclient **must** release exactly as many tokens as it has acquired from
+  the semaphore (this does not include the implicit tokens).
+
+Implementation of the protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We then propose that GHC and ``cabal``/``stack`` implement the GHC jobserver
+protocol, with GHC as a jobclient and ``cabal``/``stack` as the jobserver.
+
+As far as the GHC jobclient implementation is concerned, we have identified
+two salient points:
+
+A. There is a hidden cost in having a GHC program (such as ``ghc`` itself) run on
+   fewer CPU cores than its capabilities: the stop-the-world cost of garbage
+   collection becomes much more expensive. If we were to give GHC a fixed number
+   ``N`` of capabilities, then parallel garbage collections in that GHC would
+   recruit ``N`` OS threads. Instead we propose that GHC will adjust its number
+   of capabilities, via ``setNumCapabilities`` [8]_ , to the number of modules
+   it is compiling.
+   Note that GHC already calls ``setNumCapabilities`` with the argument to ``-j``.
+
+   We have concerns that rapidly adjusting the number of capabilities may have
+   adverse effects: we propose that GHC will limit the rate at which it can
+   release tokens to the semaphore so as to reduce this thrashing.
+   The exact mechanism of rate limiting is unspecified here.
+
+B. We believe that GHC invocations should not be too eager to release
+   semaphore tokens. This skews the balance in favour of in-unit parallelism
```

Comment:
At this point I'm still quite unclear on the algorithm for GHC requesting tokens from the job-server. How does GHC decide how many tokens to request in the first place?

Reply:
With the current implementation GHC requests 1 per module.. but this can change in future so we have left it deliberately vague.